db-scheduler-mongo/src/main/java/com/github/kagkarlsson/scheduler/MongoTaskRepository.java

```diff
@@ -0,0 +1,434 @@
+    }
+
+    private Bson buildFilterFromParams(String taskName, String taskInstance, Long version,
+        boolean withVersion) {
+        final Bson taskNameCondition = eq(Fields.taskName, taskName);
+        final Bson taskInstanceCondition = eq(Fields.taskInstance, taskInstance);
+
+        List<Bson> filters = new ArrayList<>();
+        filters.add(taskNameCondition);
+        filters.add(taskInstanceCondition);
+        if (withVersion) {
+            final Bson taskVersion = eq(Fields.version, version);
+            filters.add(taskVersion);
+        }
+
+        return and(filters);
+    }
+
+    @Override
+    public boolean reschedule(Execution execution, Instant nextExecutionTime,
+        Instant lastSuccess, Instant lastFailure, int consecutiveFailures) {
+        return rescheduleInternal(execution, nextExecutionTime, null, lastSuccess, lastFailure,
+            consecutiveFailures);
+    }
+
+    @Override
+    public boolean reschedule(Execution execution, Instant nextExecutionTime, Object newData,
+        Instant lastSuccess, Instant lastFailure, int consecutiveFailures) {
+        return rescheduleInternal(execution, nextExecutionTime, newData, lastSuccess,
+            lastFailure, consecutiveFailures);
+
+    }
+
+    private boolean rescheduleInternal(Execution execution, Instant nextExecutionTime,
+        Object data, Instant lastSuccess, Instant lastFailure, int consecutiveFailures) {
+
+        final FindOneAndUpdateOptions options = new FindOneAndUpdateOptions();
+        options.upsert(false);
+
+        Bson update = combine(set(Fields.executionTime, nextExecutionTime),
+            set(Fields.taskData, serializer.serialize(data)),
+            set(Fields.lastSuccess, lastSuccess), set(Fields.lastFailure, lastFailure),
+            set(Fields.consecutiveFailures, consecutiveFailures),
+            inc(Fields.version, 1));
+
+        final TaskEntity document = this.collection
+            .findOneAndUpdate(buildFilterFromExecution(execution), update, options);
+
+        return !Objects.isNull(document);
+    }
+
+    @Override
+    public Optional<Execution> pick(Execution e, Instant timePicked) {
+        // update where picked is true, task_name, task_instance, version
+        final Bson filterNameInstanceVersion = buildFilterFromExecution(e);
+        final Bson filters = and(filterNameInstanceVersion, eq(Fields.picked, false));
+
+        final FindOneAndUpdateOptions options = new FindOneAndUpdateOptions();
+        options.returnDocument(ReturnDocument.AFTER);
+        options.upsert(false);
+
+        Bson update = combine(set(Fields.picked, true),
+            set(Fields.pickedBy, StringUtils.truncate(schedulerSchedulerName.getName(), 50)),
+            set(Fields.lastHeartbeat, timePicked),
+            inc(Fields.version, 1));
+
+        TaskEntity document = this.collection.findOneAndUpdate(filters, update, options);
+
+        return toExecution(document);
+    }
```

Comment:
Does this method guarantee that for concurrent pick-operations, only one update will succeed? As far as I could tell it does, since it seem to have the version-filter and increments version on update. How does Mongo handle concurrent updates for a "row"? Only one suceeds?

Reply:
I have tested it, and Mongo can indeed make sure that only one thread can find the "row" that is (picked = false, version = n) and update it to (picked = true, version = n + 1), so that other threads won't fall in the same condition.
It is also describe in the example https://docs.mongodb.com/manual/tutorial/model-data-for-atomic-operations/.